fabric_generator/utilities.py

```diff
@@ -0,0 +1,1088 @@
+        after_right_index = port[(right_index+1):]
+        ExpandList = []
+        ExpandList = re.split('\|', port[left_index+1:right_index])
+        for entry in ExpandList:
+            ExpandListItem = (before_left_index+entry+after_right_index)
+            ExpandListPorts(ExpandListItem, PortList)
+
+    else:
+        # print('DEBUG: else, just:',port)
+        PortList.append(port)
+    return
+
+
+def takes_list(a_string, a_list):
+    print('first debug (a_list):', a_list, 'string:', a_string)
+    for item in a_list:
+        print('hello debug:', item, 'string:', a_string)
+
+
+def GetVerilogDeclarationForFile(VHDL_file_name):
+    ConfigPortUsed = 0  # 1 means is used
+    VHDLfile = [line.rstrip('\n')
+                for line in open(f"{src_dir}/{VHDL_file_name}")]
+    templist = []
+    # for item in VHDLfile:
+    # print(item)
+    for line in VHDLfile:
+        # NumberOfConfigBits:0 means no configuration port
+        if re.search('NumberOfConfigBits', line, flags=re.IGNORECASE):
+            # NumberOfConfigBits appears, so we may have a config port
+            ConfigPortUsed = 1
+            # but only if the following is not true
+            if re.search('NumberOfConfigBits:0', line, flags=re.IGNORECASE):
+                ConfigPortUsed = 0
+    # print('', file=file)
+    return ConfigPortUsed
+
+
+# This class represents individual tiles in the architecture
+class TileModelGen:
+    tileType = ""
+    bels = []
+    belsWithIO = []  # Currently the plan is to deprecate bels and replace it with this. However, this would require nextpnr model generation changes, so I won't do that until the VPR foundations are established
+    # Format for belsWithIO is [bel name, prefix, inputs, outputs, whether it has a clock input]
+    # Format for bels is [bel name, prefix, ports, whether it has a clock input]
+    wires = []
+    # For storing single wires (to handle cascading and termination)
+    atomicWires = []
+    pips = []
+    belPorts = set()
+    matrixFileName = ""
+    pipMuxes_MapSourceToSinks = []
+    pipMuxes_MapSinkToSources = []
+    x = -1  # Init with negative values to ease debugging
+    y = -1
+
+    def __init__(self, inType):
+        self.tileType = inType
+
+    def genTileLoc(self, separate=False):
+        if (separate):
+            return("X" + str(self.x), "Y" + str(self.y))
+        return "X" + str(self.x) + "Y" + str(self.y)
+
+
+# This class represents the fabric as a whole
+class FabricModelGen:
+    tiles = []
+    height = 0
+    width = 0
+    cellTypes = []
+
+    def __init__(self, inHeight, inWidth):
+        self.width = inWidth
+        self.height = inHeight
+
+    def getTileByCoords(self, x: int, y: int):
+        for row in self.tiles:
+            for tile in row:
+                if tile.x == x and tile.y == y:
+                    return tile
+        raise ValueError(f"{x}, {y} is not a valid tile coordinate")
```

Comment:
if you look at the old code, it returned `None` instead of an error in this case: https://github.com/gatecat/FABulous/blob/4fe503fe97fdbdbd9374c082a8348a3b9a6dd7c5/fabric_generator/fabric_gen.py#L4257

that was actually important because of what happens when you go off the edge of the fabric (which should be dealt with already by other cases so the `getTileByCoords` value isn't actually used in those cases...)